store/auth.go

```diff
@@ -80,45 +82,112 @@ func loginCaveatID(m *macaroon.Macaroon) (string, error) {
 	return caveatID, nil
 }
 
+// retryPostRequestDecodeJSON calls retryPostRequest and decodes the response into either success or failure.
+func retryPostRequestDecodeJSON(endpoint string, headers map[string]string, data []byte, success interface{}, failure interface{}) (resp *http.Response, err error) {
+	return retryPostRequest(endpoint, headers, data, func(ok bool, resp *http.Response) error {
+		result := success
+		if !ok {
+			result = failure
+		}
+		if result != nil {
+			return json.NewDecoder(resp.Body).Decode(result)
+		}
+		return nil
+	})
+}
+
+// retryPostRequest calls doRequest and decodes the response in a retry loop.
+func retryPostRequest(endpoint string, headers map[string]string, data []byte, decode func(ok bool, resp *http.Response) error) (resp *http.Response, err error) {
+	var attempt *retry.Attempt
+	startTime := time.Now()
+	for attempt = retry.Start(defaultRetryStrategy, nil); attempt.Next(); {
+		var body io.Reader
+		if data != nil {
```

Comment:
no need to do this check,

body := bytes.NewBuffer(data) would be ok (also for nil)